controllers/nutanixmachine_controller_test.go

```diff
@@ -470,6 +471,393 @@ func TestNutanixMachineReconciler_SetupWithManager_ClusterToTypedObjectsMapperEr
+							AdapterType: infrav1.NutanixMachineDiskAdapterTypeSCSI,
+						},
+						StorageConfig: &infrav1.NutanixMachineVMStorageConfig{
+							DiskMode: infrav1.NutanixMachineDiskModeStandard,
+							StorageContainer: &infrav1.NutanixResourceIdentifier{
+								UUID: ptr.To("06b1ce03-f384-4488-9ba1-ae17ebcf1f91"),
+								Type: infrav1.NutanixIdentifierUUID,
+							},
+						},
+					},
+				}
+			},
+			description: "Verify NutanixMachine with data disk device type error",
+			stepDesc:    "should error on validation due to device type",
+			errCheck: func(g *WithT, err error) {
+				g.Expect(err).To(HaveOccurred())
+				g.Expect(err.Error()).To(ContainSubstring("invalid device type not-disk for data disk"))
+			},
+		},
+		{
+			name: "dataDiskErrorWrongAdapterType",
+			dataDisks: func() []infrav1.NutanixMachineVMDisk {
+				return []infrav1.NutanixMachineVMDisk{
+					{
+						DiskSize: resource.MustParse("20Gi"),
+						DeviceProperties: &infrav1.NutanixMachineVMDiskDeviceProperties{
+							DeviceType:  infrav1.NutanixMachineDiskDeviceTypeDisk,
+							AdapterType: "not-scsi",
+						},
+						StorageConfig: &infrav1.NutanixMachineVMStorageConfig{
+							DiskMode: infrav1.NutanixMachineDiskModeStandard,
+							StorageContainer: &infrav1.NutanixResourceIdentifier{
+								UUID: ptr.To("06b1ce03-f384-4488-9ba1-ae17ebcf1f91"),
+								Type: infrav1.NutanixIdentifierUUID,
+							},
+						},
+					},
+				}
+			},
+			description: "Verify NutanixMachine with data disk adapter type error",
+			stepDesc:    "should error on validation due to adapter type",
+			errCheck: func(g *WithT, err error) {
+				g.Expect(err).To(HaveOccurred())
+				g.Expect(err.Error()).To(ContainSubstring("invalid adapter type not-scsi for data disk"))
+			},
+		},
+		{
+			name: "dataDiskErrorWrongDeviceIndex",
+			dataDisks: func() []infrav1.NutanixMachineVMDisk {
+				return []infrav1.NutanixMachineVMDisk{
+					{
+						DiskSize: resource.MustParse("20Gi"),
+						DeviceProperties: &infrav1.NutanixMachineVMDiskDeviceProperties{
+							DeviceType:  infrav1.NutanixMachineDiskDeviceTypeDisk,
+							AdapterType: infrav1.NutanixMachineDiskAdapterTypeSCSI,
+							DeviceIndex: -1,
+						},
+						StorageConfig: &infrav1.NutanixMachineVMStorageConfig{
+							DiskMode: infrav1.NutanixMachineDiskModeStandard,
+							StorageContainer: &infrav1.NutanixResourceIdentifier{
+								UUID: ptr.To("06b1ce03-f384-4488-9ba1-ae17ebcf1f91"),
+								Type: infrav1.NutanixIdentifierUUID,
+							},
+						},
+					},
+				}
+			},
+			description: "Verify NutanixMachine with data disk device index error",
+			stepDesc:    "should error on validation due to device index",
+			errCheck: func(g *WithT, err error) {
+				g.Expect(err).To(HaveOccurred())
+				g.Expect(err.Error()).To(ContainSubstring("invalid device index -1 for data disk"))
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		g := NewWithT(t)
+
+		Describe("NutanixMachineValidateDataDisks", func() {
```

Comment:
nit: I would prefer a pure go unit test, but we can refactor this in a future PR.

Reply:
I agree that it's better to get rid of ginkgo in unit tests. But I prefer to do it in future.